lib/datadog/core/telemetry/worker.rb

```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
+
+require_relative 'event'
+
+require_relative '../utils/only_once_successful'
+require_relative '../workers/polling'
+require_relative '../workers/queue'
+
+module Datadog
+  module Core
+    module Telemetry
+      # Accumulates events and sends them to the API at a regular interval, including heartbeat event.
+      class Worker
+        include Core::Workers::Queue
+        include Core::Workers::Polling
+
+        DEFAULT_BUFFER_MAX_SIZE = 1000
+
+        TELEMETRY_STARTED_ONCE = Utils::OnlyOnceSuccessful.new
+
+        def initialize(
+          heartbeat_interval_seconds:,
+          emitter:,
+          dependency_collection:,
+          enabled: true,
+          shutdown_timeout: Workers::Polling::DEFAULT_SHUTDOWN_TIMEOUT,
+          buffer_size: DEFAULT_BUFFER_MAX_SIZE
+        )
+          @emitter = emitter
+          @dependency_collection = dependency_collection
+
+          # Workers::Polling settings
+          self.enabled = enabled
+          # Workers::IntervalLoop settings
+          self.loop_base_interval = heartbeat_interval_seconds
+          self.fork_policy = Core::Workers::Async::Thread::FORK_POLICY_STOP
+
+          @shutdown_timeout = shutdown_timeout
+          @buffer_size = buffer_size
+
+          self.buffer = buffer_klass.new(@buffer_size)
+        end
+
+        def start
+          return if !enabled? || forked?
+
+          # starts async worker
+          perform
+        end
+
+        def stop(force_stop = false, timeout = @shutdown_timeout)
+          buffer.close if running?
+
+          flush_events(dequeue) if work_pending?
+
+          super
+        end
+
+        def enqueue(event)
+          buffer.push(event)
+        end
+
+        def sent_started_event?
+          TELEMETRY_STARTED_ONCE.ran?
+        end
+
+        private
+
+        def perform(*events)
+          return if !enabled? || forked?
+
+          started! unless sent_started_event?
+
+          heartbeat!
+
+          flush_events(events)
+        end
+
+        def flush_events(events)
+          return if !enabled? || !sent_started_event?
+
+          Datadog.logger.debug { "Sending #{events&.count} telemetry events" }
+          (events || []).each do |event|
+            send_event(event)
+          end
+        end
+
+        def heartbeat!
+          return if !enabled? || !sent_started_event?
+
+          send_event(Event::AppHeartbeat.new)
+        end
+
+        def started!
+          return unless enabled?
+
+          TELEMETRY_STARTED_ONCE.run do
```

Comment:
Will this try indefinitely?

Reply:
Yes, currently it will try indefinitely: probably we need to limit it.